Explain stages in terms of the compiler currently running (take N+1) (#857)
rust-lang/rustc-dev-guide: src/building/bootstrapping.md

- The "stage (N+1) compiler" is assembled from "stage N artifacts". This process is called _uplifting_.

For example, `x.py build --stage 0` means to build with the beta `rustc`, and
`test --stage 0` isn't really meaningful (it's not running tests on your changes,
but on `beta`, so it always fails). Similarly, `doc --stage 0` means to
document using the beta `rustdoc`.

Note that this implies the stage N compiler is _not_ the same as the compiler
built by `build --stage N compiler/rustc` -- that's 'stage N artifacts'
('the compiler built by stage N').

In short, _stage 0 uses the stage0 compiler to create stage0 artifacts which
will later be uplifted to stage1_.

In each stage, two major steps are performed:

1. `std` is compiled by the stage N compiler.
2. That `std` is linked to programs built by the stage N compiler, including the stage (N+1) compiler.

This is somewhat intuitive if one thinks of the stage (N+1) compiler as "just"
another program we are building with the stage N compiler:
`build --stage N compiler/rustc` is linking the stage (N+1) compiler to the `std`
built by the stage N compiler.

> Here is a chart of a full build using x.py:

Reviewer: ```suggestion
Here is a chart of a full build using `x.py`:
```

This is also where `--keep-stage 1 library/std` comes into play. Since most
changes to the compiler don't actually change the ABI, once you've produced a
`std` in stage 1, you can probably just reuse it with a different compiler.
If the ABI hasn't changed, you're good to go, no need to spend the time
recompiling that `std`.
`--keep-stage` simply assumes the previous compile is fine and copies those
artifacts into the appropriate place, skipping the cargo invocation.

### Cross-compiling

Building stage2 `std` is different depending on whether you are cross-compiling or not
(see in the table how stage2 only builds non-host `std` targets).
This is because `x.py` uses a trick: if `HOST` and `TARGET` are the same,
it will reuse stage1 `std` for stage2! This is sound because stage1 `std`
was compiled with the stage1 compiler, i.e. a compiler using the source code
you currently have checked out. So it should be identical (and therefore ABI-compatible)
to the `std` that `stage2/rustc` would compile.

However, when cross-compiling, stage1 `std` will only run on the host.
So the stage2 compiler has to recompile `std` for the target.

> ### Why does only libstd use cfg(bootstrap)?

Reviewer: ```suggestion
### Why does only libstd use `cfg(bootstrap)`?
```

## Understanding stages of bootstrap

### Overview

This is a detailed look into the separate bootstrap stages.

The convention `x.py` uses is that:
- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).
- A "stage N artifact" is an artifact that is _produced_ by the stage N compiler.
- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
  process is called _uplifting_.

For example, `x.py build --stage 0` means to build with the beta `rustc`, and
`test --stage 0` isn't really meaningful (it's not running tests on your changes,
but on `beta`, so it always fails). Similarly, `doc --stage 0` means to
document using the beta `rustdoc`.

Note that this implies the stage N compiler is _not_ the same as the compiler
built by `build --stage N compiler/rustc` -- that's 'stage N artifacts'
('the compiler built by stage N').

In short, _stage 0 uses the stage0 compiler to create stage0 artifacts which

> will later be uplifted to stage1_.

Reviewer 2: Maybe "to become the stage1 compiler" or so?

In each stage, two major steps are performed:

1. `std` is compiled by the stage N compiler.
2. That `std` is linked to programs built by the stage N compiler, including

>    the stage (N+1) compiler.

Reviewer 2: ```suggestion
   the stage N artifacts aka stage (N+1) compiler.
```

This is somewhat intuitive if one thinks of the stage (N+1) compiler as "just"
another program we are building with the stage N compiler:
`build --stage N compiler/rustc` is linking the stage (N+1) compiler to the `std`

> built by the stage N compiler.

Reviewer 2: Given the off-by-1 here everywhere, wouldn't it be better to always say "stage N artifacts" instead of "stage N+1 compiler"? That also aligns better with "stage N std" below.

Here is a chart of a full build using `x.py`:

<img alt="A diagram of the rustc compilation phases" src="../img/rustc_stages.svg" class="center" />

Keep in mind this diagram is a simplification, i.e. `rustdoc` can be built at
different stages, the process is a bit different when passing flags such as
`--keep-stage`, or if there are non-host targets.

> The stage 1 artifacts are what is shipped to end-users, including `stage2/bin/rustc`.

Reviewer 2: That is confusing since two stages appear here, wouldn't it be better to refer to `build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/rustc-main`?

Author: Hmm, the number of directories makes that a little difficult to read IMO (even if it were `build/HOST/stage1-rust/TARGET/release/rustc-main`). Maybe instead something like this?
```suggestion
The stage 1 artifacts are what is shipped to end-users. These are the same artifacts that will be uplifted to stage 2, so we sometimes say that rustup distributes stage 2.
```

Author: Or maybe it should just say 'distributes the stage 2 compiler'.

Author: I changed this to 'ships the stage 2 compiler'.

### Stages and libstd

Note that there are two `std` libraries in play here:
1. The library _linked_ to `stageN/rustc`, which was built by stage N-1 (stage N-1 `std`)
2. The library _used to compile programs_ with `stageN/rustc`, which was
   built by stage N (stage N `std`).

> stage N `std` is pretty much necessary for any useful work with the compiler.

Reviewer 2: With which compiler?

Without it, you can only compile programs with `#![no_core]` -- not terribly useful!

The reason these need to be different is because they aren't necessarily ABI-compatible:

> there could be a new layout optimization on nightly that isn't present in `beta`.

Reviewer 2: Much more common are changes to MIR or other changes to Rust metadata stored in .rlib files.

changes to the compiler don't actually change the ABI, once you've produced a
`std` in stage 1, you can probably just reuse it with a different compiler.
If the ABI hasn't changed, you're good to go, no need to spend the time
recompiling that `std`.
`--keep-stage` simply assumes the previous compile is fine and copies those
artifacts into the appropriate place, skipping the cargo invocation.

### Cross-compiling

Building stage2 `std` is different depending on whether you are cross-compiling or not
(see in the table how stage2 only builds non-host `std` targets).
This is because `x.py` uses a trick: if `HOST` and `TARGET` are the same,
it will reuse stage1 `std` for stage2! This is sound because stage1 `std`
was compiled with the stage1 compiler, i.e. a compiler using the source code
you currently have checked out. So it should be identical (and therefore ABI-compatible)
to the `std` that `stage2/rustc` would compile.

However, when cross-compiling, stage1 `std` will only run on the host.
So the stage2 compiler has to recompile `std` for the target.

### Why does only libstd use `cfg(bootstrap)`?

The `rustc` generated by the stage0 compiler is linked to the freshly-built

> `std`, which means that for the most part only `std` needs to be cfg-gated,

Reviewer 2: why "for the most part"?

Author: Not sure - @Mark-Simulacrum wrote this in https://github.com/rust-lang/rust/issues/57963#issuecomment-458809957, do you remember why?

Reviewer 3: When rustc adds internal lints that use diagnostic items, we sometimes have cfg(bootstrap) in rustc, same as lang items or diagnostic items in std. That's quite rare though.

Author: I added a mention of that, thanks!

## Understanding stages of bootstrap

### Overview

This is a detailed look into the separate bootstrap stages.

The convention `x.py` uses is that:
- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).

> - A "stage N artifact" is an artifact that is _produced_ by the stage N compiler.

Reviewer: Maybe this?

```suggestion
- A "stage N artifact" is a build that is _produced_ by the stage N compiler.
```

Author: How about 'a build artifact'? 'build' seems ambiguous to me, although I agree 'artifact' is not very clear on its own.

Reviewer: Hmm, I think "build" is clearer by itself. What does "build artifact" add?

Another option: "is a binary"

Author: It is _not_ just a binary: artifacts include shared object files and .rlibs, like the standard library. I'll add that.

Author: I think 'build' is ambiguous because it's normally used as a verb, not a noun. Saying you've built a build seems confusing.

Author: I added a section on build artifacts - does it help at all?

- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
  process is called _uplifting_.

#### Examples

- `x.py build --stage 0` means to build with the beta `rustc`.
- `x.py doc --stage 0` means to document using the beta `rustdoc`.
- `x.py test --stage 0 library/std` means to run tests on the standard library
  without building `rustc` from source ('build with stage 0, then test the artifacts').
- `x.py test --stage 0 src/test/ui` is not meaningful: it runs tests on the
  _beta_ compiler and doesn't build `rustc` from source. Use `test src/test/ui`
  instead, which builds from source.
- `x.py test --stage 0 compiler/rustc` builds the compiler but runs no tests:
  it's running `cargo test -p rustc`, but cargo doesn't understand Rust's

>   tests. You shouldn't need to use this, use `test` instead (without arguments).

Reviewer 4: Not sure this is useful. Unless I'm missing something, you shouldn't use `x.py test compiler/rustc` in the first place. So to me this paragraph hasn't anything to do with bootstrapping, and I find it kind of confusing.

Author: Hmm, I included it to show the difference between running cargo and running compiletest. Do you think people won't run it by accident?

Reviewer: I think maybe move the things you _shouldn't do_ to a different section; putting them under "examples" makes it look like you _should_ use them.

Author: Good idea, done.

### Building vs. Running

> Note that this implies the stage N compiler is _not_ the same as the compiler
> built by `build --stage N compiler/rustc` -- that's 'stage N artifacts'
> ('the compiler built by stage N').

Reviewer 4: Suggestion:
```suggestion
Note that this implies that `build --stage N compiler/rustc` does _not_ build
the stage N compiler: `--stage N` means to build 'stage N' artifacts, and
the compiler built with stage N artifacts is actually the stage (N+1) compiler.
```

Author: This is incorrect and IMO more confusing than the original. 'built with stage N artifacts' doesn't really make sense, but even if you meant 'the stage N+1 compiler' it doesn't make sense to say stage N+1 builds itself.

Maybe something like this?
```suggestion
Note that this implies that the stage N compiler is different from `build --stage N compiler/rustc`: `--stage N` means to build _using_ the stage N compiler, while `compiler/rustc` is the 'stage N artifacts', which will eventually be uplifted to stage (N+1).
```

Reviewer: What about this? It's shorter and simpler. I'm not sure if I cut out some information you want to share, but if that's true maybe have the first sentence be this and then further explain?

```suggestion
Note that `build --stage N compiler/rustc` **does not** build the stage N compiler:
instead it builds the stage _N+1_ compiler _using_ the stage N compiler.
```

I wasn't sure what "this" was referring in "Note that this implies", so I cut that part out.

Author: Yeah, I like that a lot better. Thanks!

In each stage, two major steps are performed:

1. `std` is compiled by the stage N compiler.
2. That `std` is linked to programs built by the stage N compiler, including
   the stage N artifacts (stage (N+1) compiler).

This is somewhat intuitive if one thinks of the stage N artifacts as "just"
another program we are building with the stage N compiler:
`build --stage N compiler/rustc` is linking the stage N artifacts to the `std`
built by the stage N compiler.

Here is a chart of a full build using `x.py`:

<img alt="A diagram of the rustc compilation phases" src="../img/rustc_stages.svg" class="center" />

Keep in mind this diagram is a simplification, i.e. `rustdoc` can be built at
different stages, the process is a bit different when passing flags such as
`--keep-stage`, or if there are non-host targets.

The stage 2 compiler is what is shipped to end-users.

### Stages and libstd

Note that there are two `std` libraries in play here:
1. The library _linked_ to `stageN/rustc`, which was built by stage N-1 (stage N-1 `std`)
2. The library _used to compile programs_ with `stageN/rustc`, which was
   built by stage N (stage N `std`).

> stage N `std` is pretty much necessary for any useful work with the stage N compiler.

Reviewer 4: ```suggestion
Stage N `std` is pretty much necessary for any useful work with the stage N compiler.
```

Without it, you can only compile programs with `#![no_core]` -- not terribly useful!

The reason these need to be different is because they aren't necessarily ABI-compatible:
there could be a new layout optimizations, changes to MIR, or other changes
to Rust metadata on nightly that aren't present in beta.

This is also where `--keep-stage 1 library/std` comes into play. Since most
changes to the compiler don't actually change the ABI, once you've produced a
`std` in stage 1, you can probably just reuse it with a different compiler.

> If the ABI hasn't changed, you're good to go, no need to spend the time

Reviewer 4: ```suggestion
If the ABI hasn't changed, you're good to go, no need to spend time
```

## Understanding stages of bootstrap

### Overview

This is a detailed look into the separate bootstrap stages.

The convention `x.py` uses is that:
- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).
- A "stage N artifact" is a build artifact that is _produced_ by the stage N compiler.
- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
  process is called _uplifting_.

#### Examples

- `x.py build --stage 0` means to build with the beta `rustc`.
- `x.py doc --stage 0` means to document using the beta `rustdoc`.
- `x.py test --stage 0 library/std` means to run tests on the standard library
    without building `rustc` from source ('build with stage 0, then test the
  artifacts'). If you're working on the standard library, this is normally the
  test command you want.
- `x.py test src/test/ui` means to build the stage 1 compiler and run
  `compiletest` on it. If you're working on the compiler, this is normally the
  test command you want.

> #### Non-examples

Reviewer: "Non-examples" isn't super clear:

```suggestion
#### Examples of what you _shouldn't_ do
```

## Understanding stages of bootstrap

### Overview

This is a detailed look into the separate bootstrap stages.

The convention `x.py` uses is that:
- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).
- A "stage N artifact" is a build artifact that is _produced_ by the stage N compiler.
- The "stage (N+1) compiler" is assembled from "stage N artifacts". This

>   process is called _uplifting_.

Reviewer: What do you mean when you say "assembled"?

Author: I am not actually sure ... @Mark-Simulacrum what does `Assemble` actually do? It can't just be copying files because I tried running the bin stage0-rustc and it gave an error about shared objects:
```
build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/rustc-main: error while loading shared libraries: librustc_driver-3b0ece280f85df92.so: cannot open shared object file: No such file or directory
```

#### Examples

- `x.py build --stage 0` means to build with the beta `rustc`.
- `x.py doc --stage 0` means to document using the beta `rustdoc`.
- `x.py test --stage 0 library/std` means to run tests on the standard library
    without building `rustc` from source ('build with stage 0, then test the
  artifacts'). If you're working on the standard library, this is normally the
  test command you want.
- `x.py test src/test/ui` means to build the stage 1 compiler and run
  `compiletest` on it. If you're working on the compiler, this is normally the
  test command you want.

#### Examples of what *not* to do

- `x.py test --stage 0 src/test/ui` is not meaningful: it runs tests on the
  _beta_ compiler and doesn't build `rustc` from source. Use `test src/test/ui`
  instead, which builds stage 1 from source.
- `x.py test --stage 0 compiler/rustc` builds the compiler but runs no tests:
  it's running `cargo test -p rustc`, but cargo doesn't understand Rust's
  tests. You shouldn't need to use this, use `test` instead (without arguments).
- `x.py build --stage 0 compiler/rustc` builds the compiler, but does not make

>   it usable: the build artifacts are not uplifted ([#73519]). Use `build

Reviewer: Could you explain again here what uplifting is? It's a bit unclear what exactly this means so being more verbose might be helpful.

Reviewer: Could you also show the full command for `build library/std` just to make clear what you should run?

Author: Would it be more clear if I said 'the build artifacts are not uplifted to stage1'?

Reviewer: I don't fully understand what is being said: do you mean that the artifacts for the stage 1 compiler are _built_ but are not _assembled_ (there's that word!) into the actual stage 1 compiler?

Author: Correct, that's what I'm saying. `x.py build compiler/rustc` does nothing useful.

Author: https://github.com/rust-lang/compiler-team/issues/326#issuecomment-686481525

Reviewer: I would change it to something like this then:

> `x.py build --stage 0 compiler/rustc` builds the compiler, but does not make
  it usable: the build artifacts are not assembled into the final compiler ([#73519]).

- `x.py build --stage 0` means to build with the beta `rustc`.
- `x.py doc --stage 0` means to document using the beta `rustdoc`.
- `x.py test --stage 0 library/std` means to run tests on the standard library
    without building `rustc` from source ('build with stage 0, then test the
  artifacts'). If you're working on the standard library, this is normally the
  test command you want.
- `x.py test src/test/ui` means to build the stage 1 compiler and run
  `compiletest` on it. If you're working on the compiler, this is normally the
  test command you want.

#### Examples of what *not* to do

- `x.py test --stage 0 src/test/ui` is not meaningful: it runs tests on the
  _beta_ compiler and doesn't build `rustc` from source. Use `test src/test/ui`
  instead, which builds stage 1 from source.
- `x.py test --stage 0 compiler/rustc` builds the compiler but runs no tests:
  it's running `cargo test -p rustc`, but cargo doesn't understand Rust's
  tests. You shouldn't need to use this, use `test` instead (without arguments).
- `x.py build --stage 0 compiler/rustc` builds the compiler, but does not make
  it usable: the build artifacts are not uplifted ([#73519]). Use `x.py build

>   library/std` instead which puts the compiler in `stage1/rustc`.

Reviewer: ```suggestion
  library/std` instead, which puts the compiler in `stage1/rustc`.
```

[#73519]: https://github.com/rust-lang/rust/issues/73519

> 

Reviewer: What about having a table like this? (Correct me if I got things mixed up; it's confusing :) )

I think this will help clear things up for people (me).

| Stage | Flag for _building_ this stage | Flag for _running_ this stage | Output directory | What it's used for |
| --- | --- | --- | --- | --- |
| stage 0 | N/A | `--stage 0` | N/A | This is the latest beta version of rustc, similar to `rustc +beta ...` |
| stage 1 | `--stage 1` | `--stage 1` | `build/<toolchain>/stage1` | Usually you want to build this when working on Rust |
| stage 2 | `--stage 2` | `--stage 2` | `build<toolchain>/stage2` | This is what's released to rustup; you probably don't want to build this when contributing |

```suggestion

### Stages

| Stage | Flag for _building_ this stage | Flag for _running_ this stage | Output directory | What it's used for |
| --- | --- | --- | --- | --- |
| stage 0 | N/A | `--stage 0` | N/A | This is the latest beta version of rustc, similar to `rustc +beta ...` |
| stage 1 | `--stage 1` | `--stage 1` | `build/<toolchain>/stage1` | Usually you want to build this when working on Rust |
| stage 2 | `--stage 2` | `--stage 2` | `build<toolchain>/stage2` | This is what's released to rustup; you probably don't want to build this when contributing |

```

Author: Heh, this is basically https://github.com/rust-lang/rustc-dev-guide/pull/843. Let's hold off on that for now.

Reviewer: Alright, maybe I'll open a follow-up PR then :)

Author: Also it's not entirely correct, `test --stage 1 src/test/ui` runs the stage1 compiler but `test --stage library/std` tests the stage1 _artifacts_.

Reviewer: I'm not sure what you mean... I guess that's the point :)

Author: This difference between the compiler and the standard library is basically all the confusion I've seen about staging: sometimes --stage 1 means the compiler you're _running_ and other times it means the thing you _built_ and which is which depends on a whole bunch of different things. `test compiler/rustc` is the worst offender IMO but there's a lot making it confusing.

Author: The output directories are wrong too, it should by `stage1-<component>` where `<component>` is one of

- rustc
- std
- sysroot
- tools

`build/stage1` is the compiler itself, not the artifacts.

Author: FWIW there's already a table like this here (search 'Stage 0 action'): https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html?search=#understanding-stages-of-bootstrap

Reviewer: It has a ton of detail though; I want something that's simple and gives the overview.

Author: I don't think we'll be able to get a simple chart; `--stage` means too many different things.

Reviewer: Well, maybe not "simple" but more of an overview, not the details of what exactly is happening.

Author: I can add a link to 
![image](https://user-images.githubusercontent.com/23638587/94874258-924dc680-041f-11eb-8c90-b23948b513b3.png)
but you said yourself it confused you more than it helps. I don't know any better way to say it; any view that doesn't have different sections for the compiler and standard library is going to be incomplete and misleading.

Reviewer 5: This table with the arrows is really helpful.
Could you add it to the `### Overview` section?

Author: There were a lot of people in https://github.com/rust-lang/rustc-dev-guide/pull/843 that said it confused more than it helps. I don't have the energy to push this through to an official document, I'm happy to host it on my blog or something.

Author: Or at least, if this is added I would like it to not be in this PR. This PR is mostly uncontroversial changes and gets rid of the wall of text and I don't want to lose those in a debate over what the model of bootstrapping should be.

Reviewer 5: Oh, I haven't read #843. Nevermind then.

### Building vs. Running

Note that `build --stage N compiler/rustc` **does not** build the stage N compiler:
instead it builds the stage _N+1_ compiler _using_ the stage N compiler.

In short, _stage 0 uses the stage0 compiler to create stage0 artifacts which
will later be uplifted to be the stage1 compiler_.

In each stage, two major steps are performed:

1. `std` is compiled by the stage N compiler.
2. That `std` is linked to programs built by the stage N compiler, including
   the stage N artifacts (stage (N+1) compiler).

This is somewhat intuitive if one thinks of the stage N artifacts as "just"
another program we are building with the stage N compiler:
`build --stage N compiler/rustc` is linking the stage N artifacts to the `std`
built by the stage N compiler.

Here is a chart of a full build using `x.py`:

<img alt="A diagram of the rustc compilation phases" src="../img/rustc_stages.svg" class="center" />

Keep in mind this diagram is a simplification, i.e. `rustdoc` can be built at
different stages, the process is a bit different when passing flags such as
`--keep-stage`, or if there are non-host targets.

The stage 2 compiler is what is shipped to end-users.

> ### Stages and libstd

Reviewer: You refer to it as just `std` elsewhere:

```suggestion
### Stages and `std`
```

## Understanding stages of bootstrap

### Overview

This is a detailed look into the separate bootstrap stages.

The convention `x.py` uses is that:
- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).
- A "stage N artifact" is an artifact that is _produced_ by the stage N compiler.
- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
  process is called _uplifting_.

For example, `x.py build --stage 0` means to build with the beta `rustc`, and
`test --stage 0` isn't really meaningful (it's not running tests on your changes,

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Author: Ok I definitely said this confusingly - `test --stage 0 library/std` is meaningful, but `test --stage 0 src/test/ui` is not (and `test` includes `src/test/ui`). I'll change it.

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Author: I added both `test library/std` and `test src/test/ui` as examples.

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Author: I do feel this kind of shows that the 'run stage N' model is incomplete though - why does `test --stage 0 src/test/ui`  build the compiler but not `test --stage 0 library/std`?

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Reviewer 3: I think this is just stage 0 being wonky with building things it really doesn't need, which is a bug.

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Author: I said that wrong, sorry - I meant why does src/test/ui test the stage0 _compiler_ when library/std tests the stage0 _artifacts_.

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Author: I think this is the source of a lot of confusion actually - `build --stage 0 library/std` and `test --stage 0 library/std` refer to the same code, but `build --stage 0 compiler/rustc` and `test --stage 0 src/test/ui` do not. But then it gets even more confusing - `test --stage 0 compiler/rustc` _will_ build the compiler 😕

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Reviewer 2: > test --stage 0 compiler/rustc will build the compiler

Sure, and it should, because this is running the compiler unit tests. That's just like how `--stage0 library/std` needs to build the library to run unit tests.

But `test --stage 0 src/test/ui`  should run the UI test suite on the bootstrap compiler... which is basically pointless? It should not build the compiler, sure, but it probably should just error out for not being supported.

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Reviewer 2: > uild --stage 0 library/std and test --stage 0 library/std refer to the same code, but build --stage 0 compiler/rustc and test --stage 0 src/test/ui do not. But then it gets even more confusing - test --stage 0 compiler/rustc will build the compiler confused

To hopefully deconfuse a few things, note that `build --stage 0 compiler/rustc` and `test --stage 0 compiler/rustc` *do* refer to the same code, just like for `library/std`. Only `test src/test/ui` is different, because it *tests the compiler it is run with*, whereas for `build` and crate-level `test` targets, they *test the thing they build*.

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Author: Ok, I think I understand: `test compiler/rustc` is running `cargo test` on the rustc package, which has no unit/integration tests. But `test src/test/ui` is running `compiletest`, which uses the compiler currently running.

I still think this is confusing, but I'll add a bit more explaining it.

> but on `beta`, so it always fails). Similarly, `doc --stage 0` means to

Reviewer 2: Right. `test compiler/rustc` tests the compiler *as a regular crate*, while `test src/test/ui` does way more -- it actually needs a libstd built with the *to-be-tested* compiler and then *runs it*.

`build --stage N compiler/rustc` is linking the stage N artifacts to the `std`
built by the stage N compiler.

Here is a chart of a full build using `x.py`:

<img alt="A diagram of the rustc compilation phases" src="../img/rustc_stages.svg" class="center" />

Keep in mind this diagram is a simplification, i.e. `rustdoc` can be built at
different stages, the process is a bit different when passing flags such as
`--keep-stage`, or if there are non-host targets.

The stage 1 artifacts are what is shipped to end-users, including `stage2/bin/rustc`.

### Stages and libstd

Note that there are two `std` libraries in play here:
1. The library _linked_ to `stageN/rustc`, which was built by stage N-1 (stage N-1 `std`)
2. The library _used to compile programs_ with `stageN/rustc`, which was
   built by stage N (stage N `std`).

stage N `std` is pretty much necessary for any useful work with the stage N compiler.
Without it, you can only compile programs with `#![no_core]` -- not terribly useful!

The reason these need to be different is because they aren't necessarily ABI-compatible:
there could be a new layout optimizations, changes to MIR, or other changes

> to Rust metadata on nightly that aren't present in `beta`.

Reviewer: I don’t think you have beta in backticks anywhere else:

```suggestion
to Rust metadata on nightly that aren't present in beta.
```

recompiling that `std`.
`--keep-stage` simply assumes the previous compile is fine and copies those
artifacts into the appropriate place, skipping the cargo invocation.

### Cross-compiling

Building stage2 `std` is different depending on whether you are cross-compiling or not
(see in the table how stage2 only builds non-host `std` targets).
This is because `x.py` uses a trick: if `HOST` and `TARGET` are the same,
it will reuse stage1 `std` for stage2! This is sound because stage1 `std`
was compiled with the stage1 compiler, i.e. a compiler using the source code
you currently have checked out. So it should be identical (and therefore ABI-compatible)
to the `std` that `stage2/rustc` would compile.

However, when cross-compiling, stage1 `std` will only run on the host.
So the stage2 compiler has to recompile `std` for the target.

### Why does only libstd use `cfg(bootstrap)`?

The `rustc` generated by the stage0 compiler is linked to the freshly-built
`std`, which means that for the most part only `std` needs to be cfg-gated,
so that `rustc` can use features added to std immediately after their addition,
without need for them to get into the downloaded beta.

> Note this is different from any other rust program: stage1 `rustc`

Reviewer: ```suggestion
Note this is different from any other Rust program: stage1 `rustc`
```

- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
  process is called _uplifting_.

#### Build artifacts

> Anything you can build with `x.py` is a build artifact.

Reviewer: Since you're defining a term, it might be better with italics:

```suggestion
Anything you can build with `x.py` is a _build artifact_.
```

- `x.py test --stage 0 library/std` means to run tests on the standard library
  without building `rustc` from source ('build with stage 0, then test the artifacts').
- `x.py test --stage 0 src/test/ui` is not meaningful: it runs tests on the
  _beta_ compiler and doesn't build `rustc` from source. Use `test src/test/ui` instead, which builds from source.

> - `x.py test --stage 0 compiler/rustc` builds the compiler but runs no tests: it's running `cargo test -p rustc`, but cargo doesn't understand Rust's tests. You shouldn't need to use this, use `test` instead (without arguments).

Reviewer 2: I think there actually are some unit tests in the compiler... but very few.
I agree with "do not usually use this"; I never ran this command.
